Dataset definition

ZoeMZou opened the pull request.

venexia commented on analysis/create_project_actions.R:
```diff
@@ -0,0 +1,148 @@
+
+# Return number of actions -----------------------------------------------------
+
+print(paste0("YAML created with ",length(actions_list)," actions."))
```
It looks like we made a mistake with this previously as it is saying 15 actions when you have 5 actions. The following code should fix it:

```
# Return number of actions -----------------------------------------------------

count_run_elements <- function(x) {

  if (!is.list(x)) {
    return(0)
  }

  # Check if any names of this list are "run"
  current_count <- sum(names(x) == "run", na.rm = TRUE)
  
  # Recursively check all elements in the list
  return(current_count + sum(sapply(x, count_run_elements)))

}

print(paste0("YAML created with ",count_run_elements(actions_list)," actions."))
```

venexia commented on README.md:
```diff
@@ -1,21 +1,18 @@
-The OpenSAFELY framework is a Trusted Research Environment (TRE) for electronic
-health records research in the NHS, with a focus on public accountability and
-research quality.
+-   The [`lib/`](./lib) directory contains a list of active analyses.
```
Specify which file as we will add more stuff to the lib directory as we go.

ZoeMZou replied:
Done: 
The [active_analyses](lib/active_analyses.rds) contains a list of active analyses.

venexia commented on README.md:
```diff
@@ -1,21 +1,18 @@
 # post-covid-respiratory
```
All the removed content from this file needs to be put back in.

ZoeMZou replied:
Done

venexia commented on README.md:
```diff
@@ -1,21 +1,18 @@
-# About the OpenSAFELY framework
+    -   If you are interested in how we defined our variables, we use the variable script [variable_helper_fuctions](analysis/variable_helper_functions.py) to define functions that generate variables. We then apply these functions in [variables_cohorts](analysis/variables_cohorts.py) to create a dictionary of variables for cohort definitions, and in [variables_dates](analysis/variables_dates.py)  ...
+    -   If you are interested in how we defined study dates (e.g., index and end dates), these vary by cohort and are described in the protocol. We use the script [dataset_definition_dates](analysis/dataset_definition_dates.py) to generate a dataset with all required dates for each cohort. This script imported all variables generated from [variables_dates](analysis/variables_dates.py).
+    -   If you are interested in how we defined cohorts, we use dataset definition script [dataset_definition_cohorts](analysis/dataset_definition_cohorts.py) to define the function that generate cohort. This script imported all variables generated from [variables_cohorts](analysis/variables_cohorts.py). We extracted the respective index dates and end dates for each cohort and applied this functi ...
```
Minor typo corrections below:

- If you are interested in how we defined our cohorts, we use the dataset definition script [dataset_definition_cohorts](analysis/dataset_definition_cohorts.py) to define a function that generates cohorts. This script imports all variables generated from [variables_cohorts](analysis/variables_cohorts.py) using the cohort index and end dates. This approach is used to generate three cohorts: pre-vaccination, vaccinated, and unvaccinated—found in [dataset_definition_prevax](analysis/dataset_definition_prevax.py), [dataset_definition_vax](analysis/dataset_definition_vax.py), and [dataset_definition_unvax](analysis/dataset_definition_unvax.py) respectively.

venexia replied:
I am not sure what combined is referring to in this sentence: 'Extracted data is then combined to create our final cohorts, in the [preprocess data script](analysis/preprocess_data.R).'?

ZoeMZou replied:
Directly copied from the mental health repo . Indeed nothing to combine for the current ehrQL repo. 
I will just say:
The extracted data is initially processed in the preprocess data script [preprocess data script](analysis/preprocess_data.R) to clean consultation rates, create a categorical variable for severe COVID-19, generate a flag variable for pre-existing respiratory conditions, and restrict columns, resulting in the final cohort.

venexia replied:
Some futher edits below:

> -   If you are interested in how we defined our cohorts, we use the dataset definition script [dataset_definition_cohorts](analysis/dataset_definition_cohorts.py) to define a function that generates cohorts. This script imports all variables generated from [variables_cohorts](analysis/variables_cohorts.py) using the patient's index date, the cohort start date and the cohort end date. This approach is used to generate three cohorts: pre-vaccination, vaccinated, and unvaccinated—found in [dataset_definition_prevax](analysis/dataset_definition_prevax.py), [dataset_definition_vax](analysis/dataset_definition_vax.py), and [dataset_definition_unvax](analysis/dataset_definition_unvax.py), respectively. For each cohort, the extracted data is initially processed in the [preprocess data script](analysis/preprocess_data.R), which cleans consultation rate, creates a categorical variable for severe COVID-19, generates a flag variable for pre-existing respiratory conditions, and restricts the data to relevant variables.

venexia commented on analysis/codelists.py:
```diff
@@ -0,0 +1,451 @@
+    column="code"
+)
+
+# For JCVI groups--------------------------------------------------------------------------
```
Comment formatting throughout the script could be made more consistent but looks good otherwise.

ZoeMZou replied:
Done

venexia commented on analysis/dataset_definition_cohorts.py:
I think we need to change the name of this file to better describe what it does.

ZoeMZou replied:
According to our discussion, we will leave it for now.

venexia commented on analysis/dataset_definition_dates.py:
I think we need to change the name of this file to better describe what it does.

ZoeMZou replied:
According to our discussion, no need to change as it generated a dataset of dates as suggested in the name.

venexia commented on analysis/dataset_definition_prevax.py:
```diff
@@ -0,0 +1,22 @@
+    end_prevax_outcome = Series(date)
+
+index_date = index_dates.index_prevax
+end_date_exp = index_dates.end_prevax_exposure
```
I think these need to be `end_date_exposure` and `end_date_outcome` as they are used throughout the rest of the code base like this. [This will apply to vax and unvax too.]

venexia commented on analysis/variables_cohorts.py:
```diff
@@ -0,0 +1,584 @@
+        .ctv3_code.to_category(smoking_clear)
+        )
+
+    tmp_ever_smoked = ever_matching_event_clinical_ctv3_before(ever_current_smoke, index_date)   # uses a different codelist with ONLY smoking codes
```
What does this comment mean by 'ONLY smoking codes'?

ZoeMZou replied:
https://github.com/opensafely/early-inflammatory-arthritis/blob/069e61712fcc9a0c2ec2804ff36a9b773073291c/analysis/dataset_definition.py#L149-L160u

useful codes from @milanwiedemann

venexia replied:
Okay - do we know how this differs from what we were doing in terms of how many people change classification? I am conscious that we are trying to replicate at the moment so want to be making minimal number of changes.

ZoeMZou replied:
I think they are the equivalent codes to those used in previous Cohort Extractor projects:
https://github.com/opensafely/post-covid-mentalhealth/blob/ddb50bd8fc9e4e1a9855dc9f9a03991487c908aa/analysis/common_variables.py#L617-L630

venexia replied:
Sorry, Zoe - I don't follow - if the code lists are identical, why did you swap them? And how many people change classification as a result of correcting the "ever smoke" category?

ZoeMZou replied:
Apologies for the confusion. The functionality remains identical, but the previous code required manually extracting the codelist for "ever smoke." In the updated version, I replaced it with the `filter_codes_by_category()` command, which I learned from Milan's code. This makes codes shorter.

venexia commented on analysis/variables_cohorts.py:
```diff
@@ -0,0 +1,584 @@
+        .appointment_date
+    )
+
+    ## prepare for smoking status
```
Why are these smoking codes under COVID-19 severity?

ZoeMZou replied:
We cannot use newly generated variables for calculations within the dictionary itself, so these variables need to be generated beforehand. While there isn’t strict logic required prior to defining the dictionary, incorporating these calculations within the dictionary structure is more reasonable and organized.

venexia replied:
I think the comments just need to be tidied up. You have:
1. Covid-19 prior to study start date (2020-01-01)
2. Covid-19 severity, including smoking codes

Please can you add a smoking heading?

ZoeMZou replied:
I seee, got it! I revised the heading for smoking: 
    ## Define individual temporary variables (subgroup variables) first before using them in the dictrionary
    ## 1. Covid-19 prior to study start date (2020-01-01)
    ## 2. Covid-19 severity
    ## 3. Prepare for smoking status

venexia commented on analysis/variables_cohorts.py:
```diff
@@ -0,0 +1,584 @@
+
+
+
+# Outcomes---------------------------------------------------------------------------------------------------
```
The outcomes are currently defined in primary care only as stated in the protocol. We usually also define outcomes in secondary care and death records - @rdenholm, do you know why the protocol says primary care only?

venexia replied:
Decision taken 03/12/2024 to include outcomes from primary care, secondary care and deaths as per other protocols. Recorded in issue https://github.com/opensafely/post-covid-respiratory/issues/45 and to be resolved in separate PR.

venexia commented on analysis/variables_cohorts.py:
```diff
@@ -0,0 +1,584 @@
+        .date
+        ),
+
+# DEFINE EXISTING RESPIRATORY CONDITION COHORT --------------------------------------------------------------
```
Existing respiratory conditions may also need to be defined in primary and secondary care, depending on the decision about outcomes.

venexia replied:
Decision taken 03/12/2024 to include outcomes from primary care, secondary care and deaths as per other protocols. Recorded in issue https://github.com/opensafely/post-covid-respiratory/issues/45 and to be resolved in separate PR.

venexia commented on analysis/variables_cohorts.py:
```diff
@@ -0,0 +1,584 @@
+
+        ## Region
+        cov_cat_region=practice_registrations.for_patient_on(index_date).practice_nuts1_region_name,
+        ## Consultation rate (these codes can run locally but fail in GitHub action test, details see https://docs.opensafely.org/ehrql/reference/schemas/tpp/#appointments)
```
It is okay that they fail the test - OpenSAFELY are aware of why we are using this variable and are happy for us to continue using it.

venexia commented on analysis/variables_cohorts.py:
```diff
@@ -0,0 +1,584 @@
+            ).exists_for_patient())
+        ),
+
+        ## Diabetes # to add hba1c?
```
HbA1c is only required for diabetes specific project. Please remove comment.

venexia commented on analysis/variables_cohorts.py:
```diff
@@ -0,0 +1,584 @@
+            ).exists_for_patient())
+        ),
+
+        ## All stroke (included the all-stroke codelists)
```
I don't think all stroke is needed - I can't see it on the protocol and we usually use just ischaemic (expect for the cardiovascular specific project).

ZoeMZou replied:
Removed

venexia commented on analysis/variables_cohorts.py:
```diff
@@ -0,0 +1,584 @@
+# Others
+    ## History of Covid-19 Combined
+
+        tmp_sub_bin_priorcovid19_confirmed_sgss=tmp_sub_bin_priorcovid19_confirmed_sgss,
```
Can the temporary variables be defined by the code that combines them or do they need to be outside?

ZoeMZou replied:
They need to be defined outside, so they can be included in the calculation here. Same as smoke variables. https://github.com/opensafely/post-covid-respiratory/pull/40#discussion_r1848038394

venexia commented on analysis/variables_cohorts.py:
```diff
@@ -0,0 +1,584 @@
+
+    ## Registered for a minimum of 6 months prior to the study start date # line 98: https://github.com/opensafely/comparative-booster-spring2023/blob/main/analysis/dataset_definition.py 
+
+        has_follow_up_previous_6months = (practice_registrations.for_patient_on(
```
Do we need to check registration on index_date as well to confirm that they have the 6 months? I think this code tells you they were registered on the date six months before their index but not whether they remained registered.

ZoeMZou replied:
Done:
https://github.com/opensafely/post-covid-respiratory/blob/ac685d23be5d0bafe24811cf9a952878dfae5f12/analysis/variables_cohorts.py#L520-L522

venexia reviewed (changes requested):
Hi @ZoeMZou - great work, well done! I have requested a couple of changes and there are some things to discuss at the next post-covid-events meeting. Also, please can you delete the file `dataset_definition.py` in this PR as it is no longer needed. Speak tomorrow.

ZoeMZou commented on analysis/active_analyses.R:
```diff
@@ -0,0 +1,508 @@
+##Cut points 
+prevax_cuts <- "1;28;197;365;714"
+vax_unvax_cuts <- "1;28;197"
+all_covars <- paste0("cov_cat_ethnicity;cov_cat_imd;cov_num_consultation_rate;cov_bin_healthcare_worker;",
```
Something like this? I will look into the core covars across post-covid-event projects
# Define core covariates (common across projects)
core_covars <- c(
    "cov_cat_ethnicity", "cov_cat_imd", "cov_num_consultation_rate", 
    "cov_bin_healthcare_worker", "cov_cat_smoking_status", "cov_bin_carehome_status"
)

# Define project-specific covariates (specific to the current project)
respiratory_covars <- c(
    "cov_bin_obesity", "cov_bin_history_pneumonia_snomed", "cov_bin_history_asthma_snomed", 
    "cov_bin_history_pulmonary_fibrosis_snomed", "cov_bin_ami", "cov_bin_all_stroke", 
    "cov_bin_dementia_combined", "cov_bin_liver_disease", "cov_bin_chronic_kidney_disease", 
    "cov_bin_cancer", "cov_bin_hypertension", "cov_bin_diabetes"
)

# Combine both vectors into a single vector for analysis
all_covars <- paste0(c(core_covars, project_covars), collapse = ";")

ZoeMZou replied:
I have added "cov_bin_depression", "cov_bin_copd" to the protocol and this active_analyses script. In addition, I also defined the cov_bin_depression in variables_cohorts. I will leave the outcome related variables to be resolved in separate PR.

ZoeMZou commented on analysis/active_analyses.R:
```diff
@@ -0,0 +1,508 @@
+
+
+
+
```
Done

ZoeMZou commented on analysis/active_analyses.R:
```diff
@@ -0,0 +1,508 @@
+covariate_sex <- "cov_cat_sex"
+covariate_age <- "cov_num_age"
+cox_start <- "index_date"
+cox_stop <- "end_date_out"
```
Done

ZoeMZou commented on analysis/active_analyses.R:
```diff
@@ -0,0 +1,508 @@
+                  gsub("out_date_","",df$outcome), "-", 
+                  df$population)
+
+# Pre-existing populations only run the outcomes Pneumonia and Pulmonary fibrosis
```
Done:
# Remove history of asthma from covariates for the sample with no preexisiting condition
df$covariate_other <- ifelse(
  df$population == "no_preexisting",
  sapply(
    strsplit(df$covariate_other, ";"), 
    function(covars) paste0(covars[covars != "cov_bin_history_asthma_snomed"], collapse = ";")
  ),
  df$covariate_other
)

ZoeMZou commented on analysis/variables_dates.py:
```diff
@@ -0,0 +1,584 @@
+    last_matching_med_dmd_between,
+)
+
+# Define the study_dates dictionary 
```
I extracted these reference dates from the study_dates dictionary in both scripts solely for calculation purposes, without importing them into the dataset, which is why the extraction codes appear identical. Is that correct?

ZoeMZou commented on analysis/active_analyses.R:
```diff
@@ -0,0 +1,508 @@
+
+# Specify populations and cohorts --------------------------------------------------------------
+
+populations <- c("no_preexisting","preexisting")
```
Done

venexia replied:
I think we decided to go with the subgroup (rather than the population approach)...?

ZoeMZou replied:
Done

ZoeMZou commented on analysis/variables_dates.py:
```diff
@@ -0,0 +1,584 @@
+
+death_date = minimum_of(primary_care_death_date, ons_died_from_any_cause_date)
+
+# Deregistration date- no support from any practices ---------------------------------------------
```
Done:
https://github.com/opensafely/post-covid-respiratory/blob/71046d8f9cec87faa7559a088620e6f765815bf5/analysis/variables_cohorts.py#L532-L551

venexia replied:
We no longer need `tmp_dereg_date_before_index_date` as we have `has_follow_up_previous_6months`.

Also, `tmp_dereg_date_after_index_date` shouldn't be marked tmp as it is used for inclusion/exclusion. To that end, we should decide a prefix for this group of variable - perhaps 'inex_'?

ZoeMZou replied:
I suggest leaving the renaming for inclusion/exclusion variables to a separate PR, as I’ll need to adjust the preprocess.R script accordingly when processing this group of variables.

ZoeMZou commented:
Hi @venexia ,
Thank you very much for your super helpful review! I have addressed all the points, except for the outcome variables, which are still pending a decision regarding the definition of outcomes in secondary care:

[Outcomes](https://github.com/opensafely/post-covid-respiratory/pull/40#discussion_r1838335669)
[Existing respiratory conditions](https://github.com/opensafely/post-covid-respiratory/pull/40#discussion_r1838337025)

Please could you have another review, thank you very much. 

Best wishes,
Zoe

venexia commented on analysis/active_analyses.R:
```diff
@@ -0,0 +1,520 @@
+
+# Pre-existing populations only run the outcomes Pneumonia and Pulmonary fibrosis
+# Remove Pre-existing & asthma/copd rows
+df <- df[df$population != "preex_TRUE" | !df$outcome %in% c("out_date_asthma", "out_date_copd"),]
```
This could be achieved using `outcomes_runall` and `outcomes_runmain` as for other projects if `preex_TRUE` was a subgroup rather than a population.

ZoeMZou replied:
As discussed, I have added the subgroup of preex_TRUE and preex_FALSE into loop and reflect that in the analysis coloum. So the above codes were deleted: 
https://github.com/opensafely/post-covid-respiratory/blob/ed683fbe43f338b0a9e37ad18f96623b4d43d216/analysis/active_analyses.R#L79-L109

venexia commented on analysis/active_analyses.R:
```diff
@@ -0,0 +1,520 @@
+  df$population == "preex_FALSE",
+  sapply(
+    strsplit(df$covariate_other, ";"),
+    function(covars) paste0(covars[covars != "cov_bin_history_asthma_snomed"], collapse = ";")
```
History of COPD covariate will need to be removed too.

ZoeMZou replied:
In the seperate PR for considering outcomes from primary care, secondary care and deaths, I will also rename these covariates to `cov_bin_history_asthma` and `cov_bin_history_copd` .

venexia reviewed (changes requested):
Hi Zoe - this looks great, very almost there. I think the only necessary thing to resolve is to move from population > subgroup. Otherwise, I have added some minor fixes. Note: I made a commit to add an .Rproj to this repo so remember to pull before you make further changes!

ZoeMZou commented:
Hi @venexia,

Thanks for the clarification on the comments earlier today. I’ve now completed the revisions for the PR. The main updates are:

1. active_analyses.R:

- Added `preex_FALSE` and `preex_TRUE` as subgroups in the loop, allowing the deletion of the population column and reflecting it within the analysis column.

2. variables_cohorts.py:

- Generate some variables from **preprocess.R** to **variables_cohorts.py** for consistency. 

- Renamed the inclusion/exclusion variables with the prefix inex_ for consistency.

Let me know whether they are okay to merge? 

Best wishes,
Zoe

venexia reviewed (changes requested):
Hi @ZoeMZou - some more renaming (sorry!) and some clarification needed re the changes to smoking. Things should get quicker from here - just trying to get things named in a way that will help us later.

ZoeMZou commented on analysis/active_analyses.R:
```diff
@@ -0,0 +1,516 @@
+
+# Define dates ----
+
+#study_dates <- fromJSON("output/study_dates.json")
```
Copied from the previous version. Can change here by extracting from JSON.

ZoeMZou commented on analysis/active_analyses.R:
```diff
@@ -0,0 +1,516 @@
+                  df$analysis, "-", 
+                  gsub("out_date_","",df$outcome))
+
+# Remove history covariates from preex_FALSE analyses ----
```
Great idea, thanks Venexia. 
Done:

ZoeMZou commented on analysis/variables_cohorts.py:
```diff
@@ -0,0 +1,610 @@
+        inex_alive = (((patients.date_of_death.is_null()) | (patients.date_of_death.is_after(index_date))) & 
+        ((ons_deaths.date.is_null()) | (ons_deaths.date.is_after(index_date)))),
+
+        inex_died = (patients.date_of_death.is_before(index_date) | ons_deaths.date.is_before(index_date)),
```
Dont think so, I deleted the inex_died.

ZoeMZou commented on analysis/active_analyses.R:
```diff
@@ -0,0 +1,516 @@
+)
+
+## Define project-specific covariates (specific to respiratory project) ----
+respiratory_covars <- c(
```
https://github.com/opensafely/post-covid-respiratory/blob/218c59f1a0169937cf63c226df534680563a6658/analysis/active_analyses.R#L67-L76

venexia commented on analysis/variables_cohorts.py:
```diff
@@ -0,0 +1,610 @@
+
+    ## First deregistration_date on/after index date (deregistered from all supported practices)
+
+        inex_dereg_date_after_index_date= (
```
Sorry - my comment was unclear. `purpose` should be inex_ / cov_ / exp_ / out_ / qa_ / sub_ but I couldn't recall what you were using for the censoring conditions. As you have changed death date as well, I can see we hadn't come up with a prefix. Perhaps 'cens_'?

venexia reviewed (changes requested):
Hi @ZoeMZou - great, two very small final things:
1. My comment was unclear about the use of the word 'purpose' in variable names - sorry about that. Please check the reply on the thread.
2. Please can we shorten `inex_bin_follow_up_previous_6months` to `inex_bin_6m_reg`?

ZoeMZou commented:
Hi @venexia, no problem. Silly me, thought purpose_ means creating this on purpose for censoring. All revised for `cens_date_dereg` and `cens_date_death`. Renamed `inex_bin_6m_reg` as well.

venexia reviewed (approved):
Great job, Zoe. Approved - please merge when you are ready.
